subsys/caf/modules/sensor_sampler.c

```diff
@@ -0,0 +1,206 @@
+
+static enum state state;
+
+static K_THREAD_STACK_DEFINE(sample_thread_stack, SAMPLE_THREAD_STACK_SIZE);
+static struct k_thread sample_thread;
+static int64_t uptime;
+
+
+static void report_error(void)
+{
+	state = STATE_ERROR;
+	module_set_state(MODULE_STATE_ERROR);
+}
+
+static void send_sensor_event(const char *descr, struct sensor_value *data, size_t data_cnt)
+{
+	struct sensor_event *event = new_sensor_event(sizeof(float) * data_cnt);
+	float *data_ptr = sensor_event_get_data_ptr(event);
+
+	event->descr = descr;
+
+	__ASSERT_NO_MSG(sensor_event_get_data_cnt(event) == data_cnt);
+	for (size_t i = 0; i < data_cnt; i++) {
+		data_ptr[i] = sensor_value_to_double(&data[i]);
+	}
+
+	EVENT_SUBMIT(event);
+}
+
+static int sample_sensor(const struct device *dev, const struct sensor_config *sc)
+{
+	size_t data_cnt = 0;
+
+	for (size_t i = 0; i < sc->chan_cnt; i++) {
+		data_cnt += sc->chans[i].data_cnt;
+	}
+
+	int err = 0;
+	size_t data_idx = 0;
+	struct sensor_value data[data_cnt];
+
+	for (size_t i = 0; !err && (i < sc->chan_cnt); i++) {
+		const struct sampled_channel *sampled_chan = &sc->chans[i];
+
+		err = sensor_sample_fetch_chan(dev, sampled_chan->chan);
+		if (!err) {
+			err = sensor_channel_get(dev, sampled_chan->chan, &data[data_idx]);
+			data_idx += sampled_chan->data_cnt;
+		}
+	}
+
+	if (err) {
+		LOG_ERR("Sensor sampling error (err %d)", err);
+	} else {
+		send_sensor_event(sc->event_descr, data, ARRAY_SIZE(data));
+	}
+
+	return err;
+}
+
+static int sample_sensors(int *wait_ms)
+{
+	int err = 0;
+	int64_t delta = k_uptime_delta(&uptime);
+
+	*wait_ms = INT_MAX;
+
+	for (size_t i = 0; !err && (i < ARRAY_SIZE(sensor_data)); i++) {
+		sensor_data[i].remaining_wait -= delta;
+
+		if (sensor_data[i].remaining_wait <= 0) {
+			err = sample_sensor(sensor_data[i].dev, &sensor_configs[i]);
+
+			while (sensor_data[i].remaining_wait <= 0) {
+				sensor_data[i].remaining_wait += sensor_data[i].sampling_period;
+			}
+
+			/* Run sampling again if any of the sensors was sampled. */
+			*wait_ms = 0;
+		}
+
+		if (*wait_ms > sensor_data[i].remaining_wait) {
+			*wait_ms = sensor_data[i].remaining_wait;
+		}
+	}
+
+	return err;
+}
+
+static int sensor_init(void)
+{
+	int err = 0;
+	struct sensor_data *sd;
+	const struct sensor_config *sc;
+
+	for (size_t i = 0; !err && (i < ARRAY_SIZE(sensor_data)); i++) {
+		sd = &sensor_data[i];
+		sc = &sensor_configs[i];
+
+		sd->dev = device_get_binding(sc->dev_name);
+		if (!sd->dev) {
+			err = -ENXIO;
+			break;
+		}
+		sd->sampling_period = sc->sampling_period_ms;
+		sd->remaining_wait = sc->sampling_period_ms;
+	}
+
+	if (!err) {
+		state = STATE_ACTIVE;
+		module_set_state(MODULE_STATE_READY);
+		uptime = k_uptime_get();
+	} else {
+		report_error();
+	}
+
+	return err;
+}
+
+static void sample_thread_fn(void)
+{
+	int err = 0;
+	int sleep_ms;
+
+	err = sensor_init();
+
+	while (!err) {
+		err = sample_sensors(&sleep_ms);
+
+		if (sleep_ms == INT_MAX) {
+			k_sleep(K_FOREVER);
```

Comment:
you can as well assert that as sleeping forever has no point if there is no trigger to make thread up

Reply:
Ok, changed to assert. In the nearest future I will add handling sensor triggers and then there will be something to wake up the thread.